## Final Project Submission

Please fill out:
* Student name: KIPOWE KIMATUR EMMANUEL
* Student pace: PART TIME
* Scheduled project review date/time: 05/11/2023 2000HRS
* Instructor name: SAMUEL JANE
* Blog post URL:


In [1]:
# Importation of neccesary libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Reading the dataset files and loading them into dataframes as df1, df2 & df3 for 
# title basics, title ratings & movie gross csv files respectively.
df1 = pd.read_csv("my_data/imdb.title.basics.csv.gz")
df2 = pd.read_csv("my_data/imdb.title.ratings.csv.gz")
df3 = pd.read_csv("my_data/bom.movie_gross.csv.gz")

# Datasets Exploration

### Title basics dataset

In [5]:
# df1 is the dataframe for title basics dataset. 
# Viewing the first 5 rows of our dataset
df1.head()

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [6]:
# Viewing last 5 rows of the dataset
df1.tail()

tconst                                primary_title  \
146139  tt9916538                          Kuambil Lagi Hatiku   
146140  tt9916622  Rodolpho Teóphilo - O Legado de um Pioneiro   
146141  tt9916706                              Dankyavar Danka   
146142  tt9916730                                       6 Gunn   
146143  tt9916754               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
146139                          Kuambil Lagi Hatiku        2019   
146140  Rodolpho Teóphilo - O Legado de um Pioneiro        2015   
146141                              Dankyavar Danka        2013   
146142                                       6 Gunn        2017   
146143               Chico Albuquerque - Revelações        2013   

        runtime_minutes       genres  
146139            123.0        Drama  
146140              NaN  Documentary  
146141              NaN       Comedy  
146142            116.0          NaN  
146143              NaN  Documentary

In [7]:
# Getting the column names
df1.columns

Index(['tconst', 'primary_title', 'original_title', 'start_year',
       'runtime_minutes', 'genres'],
      dtype='object')

In [8]:
# Checking the number of rows and columns
df1.shape

(146144, 6)

We note that the title basics dataframe has 146144 rows and  6 columns.

In [9]:
# A summary of the title basics dataframe
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


From the summary above,the title basics dataframe is made up of float64(1), int64(1) and object(4) datatypes.We also note that there are missing values in original_title, runtime_minutes and genre columns.

### Title ratings dataset


In [10]:
# df2 is the dataframe for title ratings dataset. 
# Viewing the first 5 rows of our dataset
df2.head()

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [11]:
# Viewing last 5 rows of the dataset
df2.tail()

tconst  averagerating  numvotes
73851  tt9805820            8.1        25
73852  tt9844256            7.5        24
73853  tt9851050            4.7        14
73854  tt9886934            7.0         5
73855  tt9894098            6.3       128

In [12]:
# Getting the column names
df2.columns

Index(['tconst', 'averagerating', 'numvotes'], dtype='object')

In [13]:
# Checking the number of rows and columns
df2.shape

(73856, 3)

We note that the title ratings dataframe has 73856 rows and  3 columns.

In [14]:
# A summary of the title ratings dataframe
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


The summary above shows that the title ratings dataframe is made up of float64(1), int64(1) & object(1) datatypes with no null values for all the columns.

### Title movie_gross dataset


In [15]:
# df3 is the dataframe for movie_gross dataset. 
# Viewing the first 5 rows of our dataset
df3.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [16]:
# Viewing last 5 rows of the dataset
df3.tail()

title      studio  domestic_gross foreign_gross  \
3382                    The Quake       Magn.          6200.0           NaN   
3383  Edward II (2018 re-release)          FM          4800.0           NaN   
3384                     El Pacto        Sony          2500.0           NaN   
3385                     The Swan  Synergetic          2400.0           NaN   
3386            An Actor Prepares       Grav.          1700.0           NaN   

      year  
3382  2018  
3383  2018  
3384  2018  
3385  2018  
3386  2018

In [17]:
# Getting the column names
df3.columns

Index(['title', 'studio', 'domestic_gross', 'foreign_gross', 'year'], dtype='object')

In [18]:
# Checking the number of rows and columns
df3.shape

(3387, 5)

The movie_gross dataframe has 73856 rows and  3 columns.

In [19]:
# A summary of the title ratings dataframe
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


The movie_gross dataframe has null values on the studio, domestic_gross and foreign_gross columns.

## Merging the Datasets

In [20]:
 # Mergiging the first two dataframes df1 and df2  using the common column tconst
first_merge = pd.merge(df1, df2, on = "tconst")

In [21]:
first_merge.head()

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  averagerating  numvotes  
0        2013            175.0    Action,Crime,Drama            7.0        77  
1        2019            114.0       Biography,Drama            7.2        43  
2        2018            122.0                 Drama            6.9      4517  
3        2018              NaN          Comedy,Drama            6.1        13  
4        2017             80.0  Comedy,Drama,Fantasy            6.5       119

In [22]:
first_merge.tail()

tconst               primary_title              original_title  \
73851  tt9913084            Diabolik sono io            Diabolik sono io   
73852  tt9914286           Sokagin Çocuklari           Sokagin Çocuklari   
73853  tt9914642                   Albatross                   Albatross   
73854  tt9914942  La vida sense la Sara Amat  La vida sense la Sara Amat   
73855  tt9916160                  Drømmeland                  Drømmeland   

       start_year  runtime_minutes        genres  averagerating  numvotes  
73851        2019             75.0   Documentary            6.2         6  
73852        2019             98.0  Drama,Family            8.7       136  
73853        2017              NaN   Documentary            8.5         8  
73854        2019              NaN           NaN            6.6         5  
73855        2019             72.0   Documentary            6.5        11

In [23]:
first_merge.shape

(73856, 8)

In [24]:
df3.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [25]:
df3.tail()

title      studio  domestic_gross foreign_gross  \
3382                    The Quake       Magn.          6200.0           NaN   
3383  Edward II (2018 re-release)          FM          4800.0           NaN   
3384                     El Pacto        Sony          2500.0           NaN   
3385                     The Swan  Synergetic          2400.0           NaN   
3386            An Actor Prepares       Grav.          1700.0           NaN   

      year  
3382  2018  
3383  2018  
3384  2018  
3385  2018  
3386  2018

In [26]:
# We note that the column named title on the movie_gross(df3) dataframe carry same contents as the original_title on the first_merge data frame.
# Renaming of column named title to original_title on the df3 dataframe(movie_gross)
df3 = df3.rename(columns={'title': 'original_title'})

In [27]:
# Checking if the changes have been effected
df3.head()

original_title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [28]:
# Now that we have a common column 'original_title', We merge first_merge with df3
df = pd.merge(first_merge, df3, on = "original_title")

In [29]:
df.head()

tconst                    primary_title  \
0  tt0315642                            Wazir   
1  tt0337692                      On the Road   
2  tt4339118                      On the Road   
3  tt5647250                      On the Road   
4  tt0359950  The Secret Life of Walter Mitty   

                    original_title  start_year  runtime_minutes  \
0                            Wazir        2016            103.0   
1                      On the Road        2012            124.0   
2                      On the Road        2014             89.0   
3                      On the Road        2016            121.0   
4  The Secret Life of Walter Mitty        2013            114.0   

                    genres  averagerating  numvotes   studio  domestic_gross  \
0       Action,Crime,Drama            7.1     15378  Relbig.       1100000.0   
1  Adventure,Drama,Romance            6.1     37886      IFC        744000.0   
2                    Drama            6.0         6      IFC        744000.0   
3                    Drama            5.7       127      IFC        744000.0   
4   Adventure,Comedy,Drama            7.3    275300      Fox      58200000.0   

  foreign_gross  year  
0           NaN  2016  
1       8000000  2012  
2       8000000  2012  
3       8000000  2012  
4     129900000  2013

In [30]:
df.shape

(2447, 12)

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2447 entries, 0 to 2446
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   tconst           2447 non-null   object 
 1   primary_title    2447 non-null   object 
 2   original_title   2447 non-null   object 
 3   start_year       2447 non-null   int64  
 4   runtime_minutes  2402 non-null   float64
 5   genres           2443 non-null   object 
 6   averagerating    2447 non-null   float64
 7   numvotes         2447 non-null   int64  
 8   studio           2444 non-null   object 
 9   domestic_gross   2429 non-null   float64
 10  foreign_gross    1573 non-null   object 
 11  year             2447 non-null   int64  
dtypes: float64(3), int64(3), object(6)
memory usage: 248.5+ KB


Now this is the data that we use for the analysis, but we first need to clean it. There are missing values on runtime_minutes,genres,studio, domestic_gross and foreign_gross columns.


## Data cleaning

Now that we have merged all our three datasets, we are now head to cleaning it before we do the analysis.

In [32]:
# Getting a summary statistics of the final merged dataframe(df) columns with numeric values.
df.describe()

start_year  runtime_minutes  averagerating      numvotes  \
count  2447.000000      2402.000000    2447.000000  2.447000e+03   
mean   2013.774418       106.810575       6.406293  7.273033e+04   
std       2.496398        20.061202       1.045029  1.345874e+05   
min    2010.000000         3.000000       1.600000  5.000000e+00   
25%    2012.000000        94.000000       5.800000  3.776000e+03   
50%    2014.000000       104.000000       6.500000  2.074600e+04   
75%    2016.000000       118.000000       7.100000  8.063900e+04   
max    2019.000000       186.000000       9.200000  1.841066e+06   

       domestic_gross         year  
count    2.429000e+03  2447.000000  
mean     3.612440e+07  2014.001635  
std      6.952373e+07     2.464797  
min      1.000000e+02  2010.000000  
25%      3.020000e+05  2012.000000  
50%      5.100000e+06  2014.000000  
75%      4.300000e+07  2016.000000  
max      7.001000e+08  2018.000000

In [34]:
# Checking the number of null values for every column
null_values = df.isnull().sum()
null_values

tconst               0
primary_title        0
original_title       0
start_year           0
runtime_minutes     45
genres               4
averagerating        0
numvotes             0
studio               3
domestic_gross      18
foreign_gross      874
year                 0
dtype: int64

In [36]:
# Since the missing values in genres and studio is significaltly small, we drop them. It wont affect the analysis.
columns = ['studio', 'genres']
df.dropna(subset=columns, inplace=True)

In [39]:
null_values = df.isnull().sum()
null_values

tconst               0
primary_title        0
original_title       0
start_year           0
runtime_minutes     43
genres               0
averagerating        0
numvotes             0
studio               0
domestic_gross      16
foreign_gross      873
year                 0
dtype: int64

In [45]:
#Check the runtim_minutes mode, mean and mode to determine which to use to fill the null values.
runtime_mode = df['runtime_minutes'].mode().values[0]
runtime_median = df['runtime_minutes'].median()
runtime_mean = df['runtime_minutes'].mean()
print(runtime_mode)
print(runtime_median)
print(runtime_mean)

100.0
104.0
106.83646224447226


In [47]:
#Check the domestic_gross mode, mean and mode to determine which to use to fill the null values.
domestic_mode = df['domestic_gross'].mode().values[0]
domestic_median = df['domestic_gross'].median()
domestic_mean = df['domestic_gross'].mean()
print(domestic_mode)
print(domestic_median)
print(domestic_mean)

1300000.0
5050000.0
36115372.06394389


In [58]:
#Check the foreign_gross mode, mean and mode to determine which to use to fill the null values.
foreign_mode = df['foreign_gross'].mode().values[0]
foreign_median = df['foreign_gross'].median()
foreign_mean = df['foreign_gross'].mean()
print(foreign_mode)
print(foreign_median)
print(foreign_mean)

TypeError: 

We note that we cannot get the mean of the foreign_gross because it has ',' i.e "1,019.4" 
The ',' needs to be removed for us to be able to compute the measures of central tendency.

In [59]:
# Formulation of a function to remove the comman that can be applied on the dataset
def remove_commas(value):
    return value.replace(',', '')

In [60]:
# Apply it to our dataframe to remove commas from our foreign_gross colums
df['foreign_gross'] = df['foreign_gross'].str.replace(',', '')

In [63]:
# we Convert it to float then
# Recheck the foreign_gross mode, mean and mode
df['foreign_gross'] = df['foreign_gross'].astype(float)
foreign_mode = df['foreign_gross'].mode().values[0]
foreign_median = df['foreign_gross'].median()
foreign_mean = df['foreign_gross'].mean()
print(foreign_mode)
print(foreign_median)
print(foreign_mean)

4200000.0
23400000.0
81652031.13395022


The remove_commas funtion has worked perfectly.We now proceed with filling our missing data. I decided to us the mean till the null values.

In [64]:
df['runtime_minutes'].fillna(runtime_mean, inplace=True)
df['domestic_gross'].fillna(domestic_mean, inplace=True)
df['foreign_gross'].fillna(foreign_mean, inplace=True)

In [66]:
# Check for missing values again
null_values = df.isnull().sum()
null_values

tconst             0
primary_title      0
original_title     0
start_year         0
runtime_minutes    0
genres             0
averagerating      0
numvotes           0
studio             0
domestic_gross     0
foreign_gross      0
year               0
dtype: int64

We have filled our null values with the mean of each column. Now we are ready to analyze our data.

# Data Analysis